src/bin/ion/commands/beta/generate/utils.rs

```diff
@@ -0,0 +1,177 @@
+
+/// Represents a field that will be added to generated data model.
+/// This will be used by the template engine to fill properties of a struct/classs.
+#[derive(Serialize)]
+pub struct Field {
+    pub(crate) name: String,
+    pub(crate) value: String,
+}
+
+/// Represents an import statement in a module file.
+/// This will be used by template engine to fill import statements of a type definition.
+#[derive(Serialize)]
+pub struct Import {
+    pub(crate) module_name: String,
+    pub(crate) type_name: String,
+}
+
+/// Represent the programming language for code generation.
+#[derive(Debug, Clone, PartialEq)]
+pub enum Language {
+    Rust,
+    Java,
+}
+
+impl Language {
+    pub fn file_extension(&self) -> &str {
+        match self {
+            Language::Rust => "rs",
+            Language::Java => "java",
+        }
+    }
+
+    pub fn file_name(&self, name: &str) -> String {
+        match self {
+            Language::Rust => name.to_case(Case::Snake),
+            Language::Java => name.to_case(Case::UpperCamel),
+        }
+    }
+}
+
+impl From<&str> for Language {
+    fn from(value: &str) -> Self {
+        match value {
+            "java" => Language::Java,
+            "rust" => Language::Rust,
+            _ => unreachable!("Unsupported programming language: {}, this tool only supports Java and Rust code generation.", value)
+        }
+    }
+}
+
+impl Display for Language {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Language::Rust => {
+                write!(f, "rust")
+            }
+            Language::Java => {
+                write!(f, "java")
+            }
+        }
+    }
+}
+
+/// Represents all the supported templates for code generation.
+/// These templates will be used by [tera] templating engine to render the generated code with appropriate context value.
+/// _Note: These template variants are based on Rust programming language.
+/// [Template::name] provides an appropriate template file name based on given programming language._
+///
+/// [tera]: <https://docs.rs/tera/latest/tera/>
+pub enum Template {
+    Struct, // Represents a template for a Rust struct or Java class
+}
+
+impl Template {
+    /// Returns a string that represent the template file name based on given programming language.
+    pub fn name(&self, language: &Language) -> &str {
+        match language {
+            Language::Rust => "struct",
+            Language::Java => "class",
+        }
+    }
+}
+
+impl TryFrom<Option<&DataModel>> for Template {
+    type Error = CodeGenError;
+
+    fn try_from(value: Option<&DataModel>) -> Result<Self, Self::Error> {
+        match value {
+            Some(DataModel::Value) | Some(DataModel::Sequence) | Some(DataModel::Struct) => {
+                Ok(Template::Struct)
+            }
+            None => invalid_data_model_error(
+                "Can not get a template without determining data model first.",
+            ),
```

Comment:
There's a point in the code gen method where we should go from `Option<DataModel>` to `DataModel`. This error should happen there, and all methods downstream of that point (including this one) should operate on `DataModel` instead of `Option<DataModel>`.

Reply:
I think after the first constraint is determined or when the generation starts for each constraint we can make this `DataModel` and return error accordingly. will add issue for it.

Reply:
https://github.com/amazon-ion/ion-cli/issues/74